sklearn/cluster/_k_means.pyx

```diff
@@ -327,18 +375,24 @@ def _centers_sparse(X, np.ndarray[INT, ndim=1] labels, n_clusters,
 
     cdef np.npy_intp cluster_id
 
-    cdef np.ndarray[DOUBLE, ndim=1] data = X.data
+    cdef np.ndarray[floating, ndim=1] data = X.data
     cdef np.ndarray[int, ndim=1] indices = X.indices
     cdef np.ndarray[int, ndim=1] indptr = X.indptr
 
-    cdef np.ndarray[DOUBLE, ndim=2, mode="c"] centers = \
-        np.zeros((n_clusters, n_features))
+    cdef np.ndarray[floating, ndim=2, mode="c"] centers
     cdef np.ndarray[np.npy_intp, ndim=1] far_from_centers
     cdef np.ndarray[np.npy_intp, ndim=1, mode="c"] n_samples_in_cluster = \
         bincount(labels, minlength=n_clusters)
     cdef np.ndarray[np.npy_intp, ndim=1, mode="c"] empty_clusters = \
         np.where(n_samples_in_cluster == 0)[0]
 
+    if floating is float:
+        centers = np.zeros((n_clusters, n_features), dtype=np.float32)
+    elif floating is double:
+        centers = np.zeros((n_clusters, n_features), dtype=np.float64)
+    else:
+        raise ValueError("Unknown floating type.")
```

Comment:
Similar remark for the allocation of the `center_squared_norms` and `centers` arrays in other functions of this file.